test/test_testing.py

```diff
@@ -5485,3 +5488,199 @@ def test_bad_results(self):
+            'container': 'bar',
+            'notice_id': id3,
+            'notice_type': 'custom',
+            'notice_key': 'example.com/n1',
+        }]
+        self.assertEqual(harness.charm.changes, expected_changes)
+
+    def test_notify_no_repeat(self):
+        """Ensure event doesn't get triggered when notice occurs but doesn't repeat."""
+        harness = ops.testing.Harness(ContainerEventCharm, meta="""
+            name: notifier
+            containers:
+              foo:
+                resource: foo-image
+        """)
+        self.addCleanup(harness.cleanup)
+        harness.begin()
+        harness.charm.observe_container_events('foo')
+
+        id1a = harness.pebble_notify('foo', 'example.com/n1',
+                                     repeat_after=datetime.timedelta(days=1))
+        id1b = harness.pebble_notify('foo', 'example.com/n1',
+                                     repeat_after=datetime.timedelta(days=1))
+
+        self.assertEqual(id1a, id1b)
+
+        expected_changes = [{
+            'name': 'pebble-custom-notice',
+            'container': 'foo',
+            'notice_id': id1a,
+            'notice_type': 'custom',
+            'notice_key': 'example.com/n1',
+        }]
+        self.assertEqual(harness.charm.changes, expected_changes)
+
+    def test_notify_no_begin(self):
+        num_notices = 0
+
+        class TestCharm(ops.CharmBase):
+            def __init__(self, framework: ops.Framework):
+                super().__init__(framework)
+                self.framework.observe(self.on['c1'].pebble_custom_notice,
+                                       self._on_pebble_custom_notice)
+
+            def _on_pebble_custom_notice(self, event: ops.PebbleCustomNoticeEvent):
+                nonlocal num_notices
+                num_notices += 1
+
+        harness = ops.testing.Harness(TestCharm, meta="""
+            name: notifier
+            containers:
+              c1:
+                resource: c1-image
+        """)
+        self.addCleanup(harness.cleanup)
+
+        id = harness.pebble_notify('c1', 'example.com/n1')
+
+        self.assertIsInstance(id, str)
+        self.assertNotEqual(id, '')
+        self.assertEqual(num_notices, 0)
+
+
+class PebbleNoticesMixin:
+    client: ops.pebble.Client
+
+    assertEqual = unittest.TestCase.assertEqual  # noqa
+    assertIsNone = unittest.TestCase.assertIsNone  # noqa
+    assertLess = unittest.TestCase.assertLess  # noqa
+    assertRaises = unittest.TestCase.assertRaises  # noqa
+    assertGreaterEqual = unittest.TestCase.assertGreaterEqual  # noqa
+
+    def test_get_notice_by_id(self):
+        client = self.client
+        key1 = 'example.com/' + os.urandom(16).hex()
+        key2 = 'example.com/' + os.urandom(16).hex()
+        id1 = client.notify(pebble.NoticeType.CUSTOM, key1)
+        id2 = client.notify(pebble.NoticeType.CUSTOM, key2, data={'x': 'y'})
+        time.sleep(0.000_001)  # Ensure times are different.
```

Comment:
it is usually a lot better to have ways of injecting a time source rather than sleeping to get time to change.
probably a fair bit of overhead for this test, but eventually sleep based tests bite you on someone else's hardware acting differently than you expected.

Reply:
Agreed. I would have patched/injected, but these tests are used against "real Pebble" as well (via the mixin). So we need real time for that (unless we want to inject something in Pebble, but that gets *really* painful).